sr/robot3/kch.py

```diff
@@ -0,0 +1,210 @@
+
+    async def main(self) -> None:
+        """Main method of the command."""
+        await self.wait_loop()
+
+    @classmethod
+    def get_kch(cls) -> Optional[str]:
+        """Get metadata."""
+        kchdc = cls(False, None)
+
+        try:
+            loop.run_until_complete(asyncio.wait_for(kchdc.run(), timeout=0.1))
+            if kchdc._kch_info is not None:
+                return kchdc._kch_info
+            else:
+                return None
+        except ConnectionRefusedError:
+            LOGGER.warning("Unable to connect to MQTT broker")
+            return None
+        except asyncio.TimeoutError:
+            LOGGER.warning("KCH Daemon took too long to respond, giving up.")
+            return None
+
+
+class KCHDaemonControlConsumer(StateConsumer):
+    """KCH LED Control Consumer."""
+
+    dependencies = ["kchd"]
+
+    def _init(self) -> None:
+        self._ready = asyncio.Event()
+
+    def _setup_logging(self, verbose: bool, *, welcome_message: bool = True) -> None:
+        """Use the logging from sr-robot3."""
+        # Suppress INFO messages from gmqtt
+        logging.getLogger("gmqtt").setLevel(logging.WARNING)
+
+    @property
+    def name(self) -> str:
+        """
+        MQTT client name of the data component.
+
+        This should be unique, as clashes will cause unexpected disconnections.
+        """
+        return f"sr-robot3-kchd-{uuid4()}"
+
+    async def main(self) -> None:
+        """Main method of the command."""
+        self._ready.set()
+        await self.wait_loop()
+
+
+class KCHLEDUpdateManagerRequest(ManagerRequest):
+    """A request to change the controllable LEDs."""
+
+    start: bool = False
+    a: Tuple[bool, bool, bool] = (False, False, False)
+    b: Tuple[bool, bool, bool] = (False, False, False)
+    c: Tuple[bool, bool, bool] = (False, False, False)
+
+
+class SRKCHDaemonBackend(RGBLEDInterface, Backend):
+    """KCH Daemon Backend."""
+
+    board = KCHBoard
+
+    @classmethod
+    def discover(cls) -> Set[Board]:
+        """
+        Discover boards that this backend can control.
+
+        :returns: set of boards that this backend can control.
+        """
+        info = KCHDaemonConsumer.get_kch()
+        if info:
+            board = KCHBoard(info, cls())
+            return {cast(Board, board)}
+        else:
+            return set()
+
+    def __init__(self) -> None:
+        self._start = False
+        self._leds = {
+            i: [False, False, False]
+            for i in range(3)
+        }
+
+        self._kchdc = KCHDaemonControlConsumer(False, None)
+        try:
+            asyncio.ensure_future(self._kchdc.run())
+        except ConnectionRefusedError:
+            LOGGER.warning("Unable to connect to MQTT broker")
+        except asyncio.TimeoutError:
+            LOGGER.warning("KCH Daemon took too long to respond, giving up.")
+
+    async def _set_leds(self) -> None:
+        await self._kchdc._ready.wait()
```

Comment:
We could add a timeout on the wait so it doesn't hang indenfinitely if kchd crashes.

```suggestion
    try:
        await asyncio.wait_for(self._kchdc._ready.wait(), timeout=1.0)
    except asyncio.TimeoutError:
        LOGGER.error("Failed to set LED on KCH")
        return
```